tests/TPP.Persistence.MongoDB.Tests/Repos/BadgeMarketRepoTest.cs

```diff
@@ -0,0 +1,218 @@
+
+        public IBadgeMarketRepo CreateBadgeMarketRepo()
+        {
+            IMongoDatabase db = CreateTemporaryDatabase();
+            _mockUserRepo = new Mock<IUserRepo>();
+            _mockBank = new Mock<IBank<User>>();
+            _mockBadgeRepo = new Mock<IBadgeRepo>();
+
+            BadgeMarketRepo badgeMarketRepo = new BadgeMarketRepo(db, _mockBadgeRepo.Object, _mockUserRepo.Object, _mockBank.Object, Mock.Of<IClock>());
+            return badgeMarketRepo;
+        }
+
+        internal class MockClock : IClock
+        {
+            public Instant FixedCurrentInstant = Instant.FromUnixTimeSeconds(1234567890);
+            public Instant GetCurrentInstant() => FixedCurrentInstant;
+        }
+
+        [Test]
+        public async Task CreateBuyOffer_write_then_read_are_equal()
+        {
+            string userId = "m4";
+            PkmnSpecies species = PkmnSpecies.OfId("1");
+            string form = "";
+            Badge.BadgeSource source = Badge.BadgeSource.ManualCreation;
+            bool shiny = true;
+            int price = 999;
+            int amount = 1;
+
+            IBadgeMarketRepo badgeMarketRepo = CreateBadgeMarketRepo();
+
+            _mockBadgeRepo.Setup(r => r.FindAllForSaleByCustom(null, species, null, null, shiny)).Returns(Task.FromResult(new List<Badge>()));
+
+            BadgeBuyOffer offer = await badgeMarketRepo.CreateBuyOffer(userId, species, form, source, shiny, price, amount, Instant.MaxValue);
+
+            Assert.That(userId, Is.EqualTo(offer.UserId));
+            Assert.That(species, Is.EqualTo(offer.Species));
+            Assert.That(form, Is.EqualTo(offer.Form));
+            Assert.That(source, Is.EqualTo(offer.Source));
+            Assert.That(shiny, Is.EqualTo(offer.Shiny));
+            Assert.That(price, Is.EqualTo(offer.Price));
+            Assert.That(amount, Is.EqualTo(offer.Amount));
+        }
+
+        [Test]
+        public async Task CreateSellOffer_sells_newest_badge()
+        {
+            string userId = "thelegend27";
+            PkmnSpecies species = PkmnSpecies.OfId("1");
+            string form = "";
+            Badge.BadgeSource source = Badge.BadgeSource.ManualCreation;
+            bool shiny = true;
+            int price = 999;
+            Instant before = Instant.FromUtc(2000, 1, 1, 0, 0);
+            Instant after = before.PlusNanoseconds(69);
+
+            IBadgeMarketRepo badgeMarketRepo = CreateBadgeMarketRepo();
+
+            Badge badgeA = new Badge("A", userId, species, source, before, form, shiny);
+            Badge badgeB = new Badge("B", userId, species, source, after, form, shiny);
+            Badge badgeBForSale = new Badge("B", userId, species, source, after, form, shiny) { SellPrice = price };
+            List<Badge> notForSale = new List<Badge> { badgeA, badgeB };
+
+            _mockBadgeRepo.Setup(m => m.FindAllNotForSaleByCustom(userId, species, form, source, shiny)).Returns(Task.FromResult(notForSale));
```

Comment:
iirc there's a `ReturnsAsync` that removes the need to wrap the result in `Task.FromResult(...)`